framework/src/node/generator/generator.ts

```diff
@@ -0,0 +1,468 @@
+		}
+	}
+
+	/**
+	 * Loads transactions from the network:
+	 * - Validates each transaction from the network and applies a penalty if invalid.
+	 * - Calls processUnconfirmedTransaction for each transaction.
+	 */
+	private async _getUnconfirmedTransactionsFromNetwork(): Promise<void> {
+		this._logger.info('Loading transactions from the network');
+
+		const { data } = ((await this._network.request({
+			procedure: NETWORK_RPC_GET_TRANSACTIONS,
+		})) as unknown) as {
+			data: Buffer;
+		};
+		const encodedData = codec.decode<GetTransactionResponse>(getTransactionsResponseSchema, data);
+
+		const validatorErrors = validator.validate(getTransactionsResponseSchema, encodedData);
+		if (validatorErrors.length) {
+			throw new LiskValidationError(validatorErrors);
+		}
+
+		const transactions = encodedData.transactions.map(transaction =>
+			this._chain.dataAccess.decodeTransaction(transaction),
+		);
+
+		for (const transaction of transactions) {
+			const { error } = await this._pool.add(transaction);
+
+			if (error) {
+				this._logger.error({ err: error }, 'Failed to add transaction to pool.');
+				throw error;
+			}
+		}
+	}
+
+	private async _generateLoop(): Promise<void> {
+		const apiContext = new APIContext({
+			stateStore: new StateStore(this._blockchainDB),
+			eventQueue: new EventQueue(),
+		});
+
+		const MS_IN_A_SEC = 1000;
+		const currentTime = Math.floor(new Date().getTime() / MS_IN_A_SEC);
+		const currentSlot = this._consensus.getSlotNumber(apiContext, currentTime);
+
+		const currentSlotTime = this._consensus.getSlotTime(apiContext, currentSlot);
+
+		const { waitThreshold } = this._config;
+		const lastBlockSlot = this._consensus.getSlotNumber(
+			apiContext,
+			this._chain.lastBlock.header.timestamp,
+		);
+
+		if (currentSlot === lastBlockSlot) {
+			this._logger.trace({ slot: currentSlot }, 'Block already forged for the current slot');
+			return;
+		}
+
+		const generator = await this._consensus.getGenerator(apiContext, currentTime);
+		const validatorKeypair = this._keypairs.get(generator);
+
+		if (validatorKeypair === undefined) {
+			this._logger.debug({ currentSlot }, 'Waiting for delegate slot');
+			return;
+		}
+
+		// If last block slot is way back than one block
+		// and still time left as per threshold specified
+		if (lastBlockSlot < currentSlot - 1 && currentTime <= currentSlotTime + waitThreshold) {
+			this._logger.info('Skipping forging to wait for last block');
+			this._logger.debug(
+				{
+					currentSlot,
+					lastBlockSlot,
+					waitThreshold,
+				},
+				'Slot information',
+			);
+			return;
+		}
+		const generatedBlock = await this._generateBlock(generator, validatorKeypair, currentTime);
+		this._logger.info(
+			{
+				id: generatedBlock.header.id,
+				height: generatedBlock.header.height,
+				generatorAddress: generator.toString('hex'),
+			},
+			'Generated new block',
+		);
+
+		await this._consensus.execute(generatedBlock as never);
+	}
+
+	private async _generateBlock(
```

Comment:
this is new logic for block generation to use state machine and module hooks (init and seal)